### PR TITLE
User authentication

### PR DESCRIPTION
**Pull request by Salisbum**
Added sign up for users.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+RSpec.describe User, type: :model do
+  let(:user) do
+    User.new({
+    email: "derpson@email.com",
+    password: "12345678"
+    })
+  end
+
+describe ".new" do
```
Comment (Salisbum) [inline spec/models/user_spec.rb]: fixed

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  let(:user) do
+    User.new({
+    email: "derpson@email.com",
+    password: "12345678"
+    })
```
Comment (Salisbum) [inline spec/models/user_spec.rb]: I feel like this is unnecessary, we're following convention as used by LA.

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  let(:user) do
+    User.new({
+    email: "derpson@email.com",
```
Comment (Salisbum) [inline spec/models/user_spec.rb]: I feel like this is unnecessary, we're following convention as used by LA.

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  let(:user) do
+    User.new({
```
Comment (Salisbum) [inline spec/models/user_spec.rb]: This doesn't make sense.

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    email: "derpson@email.com",
+    password: "12345678"
+    })
+  end
+
+describe ".new" do
+    it "should be an User object" do
+      expect(user).to be_a(User)
+    end
+  end
```
Comment (Salisbum) [inline spec/models/user_spec.rb]: fixed.

File: db/migrate/20160405143850_devise_create_users.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
+      # t.string   :unlock_token # Only if unlock strategy is :email or :both
+      # t.datetime :locked_at
+
+
```
Comment (Salisbum) [inline db/migrate/20160405143850_devise_create_users.rb]: Not supposed to edit a migration?

File: db/migrate/20160405164120_add_confirmableto_users.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+class AddConfirmabletoUsers < ActiveRecord::Migration
+  def self.up
+    add_column :users, :confirmation_token,   :string
+    add_column :users, :confirmed_at,         :datetime
+    add_column :users, :confirmation_sent_at, :datetime
+    add_column :users, :unconfirmed_email,    :string
+
+    add_index  :users, :confirmation_token, :unique => true
```
Comment (Salisbum) [inline db/migrate/20160405164120_add_confirmableto_users.rb]: Again, migration.

File: db/migrate/20160405143850_devise_create_users.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      t.inet     :last_sign_in_ip
+
+      ## Confirmable
+      # t.string   :confirmation_token
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
```
Comment (Salisbum) [inline db/migrate/20160405143850_devise_create_users.rb]: Migration.

File: config/initializers/devise.rb [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+  # :sha1, :sha512 or encryptors from others authentication tools as :clearance_sha1,
+  # :authlogic_sha512 (then you should set stretches above to 20 for default behavior)
+  # and :restful_authentication_sha1 (then you should set stretches to 10, and copy
+  # REST_AUTH_SITE_KEY to pepper).
+  #
+  # Require the `devise-encryptable` gem when using anything other than bcrypt
+  # config.encryptor = :sha512
+
+  # ==> Scopes configuration
+  # Turn scoped views on. Before rendering "sessions/new", it will first check for
```
Comment (Salisbum) [inline config/initializers/devise.rb]: is a comment? 2 characters.

File: config/initializers/devise.rb [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+  config.sign_out_via = :delete
+
+  # ==> OmniAuth
+  # Add a new OmniAuth provider. Check the wiki for more information on setting
+  # up on your models and hooks.
+  # config.omniauth :github, 'APP_ID', 'APP_SECRET', scope: 'user,public_repo'
+
+  # ==> Warden configuration
+  # If you want to use other strategies, that are not supported by Devise, or
+  # change the failure app, you can configure them inside the config.warden block.
```
Comment (Salisbum) [inline config/initializers/devise.rb]: Comment, useful info, 2 characters.

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -7,6 +7,26 @@
   <%= csrf_meta_tags %>
 </head>
 <body>
+<% if user_signed_in? %>
+  <li>
+  <%= link_to('Logout', destroy_user_session_path, :method => :delete) %>
```
Comment (smkopp92) [inline app/views/layouts/application.html.erb]: This line should be indented 1 tab further.

File: app/views/layouts/application.html.erb [hunk 12]
```diff
@@ -7,6 +7,26 @@
...
+  </li>
+<% else %>
+  <li>
+  <%= link_to('Login', new_user_session_path)  %>
```
Comment (smkopp92) [inline app/views/layouts/application.html.erb]: And Here

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -7,6 +7,26 @@
...
+  </li>
+<% end %>
+
+<% if user_signed_in? %>
+  <li>
+  <%= link_to('Edit registration', edit_user_registration_path) %>
```
Comment (smkopp92) [inline app/views/layouts/application.html.erb]: And yet again!

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -7,6 +7,26 @@
...
+  </li>
+<% else %>
+  <li>
+  <%= link_to('Sign Up', new_user_registration_path)  %>
```
Comment (smkopp92) [inline app/views/layouts/application.html.erb]: Once more!

**Comment by Salisbum** [comment]
RE: the Hound errors not marked as fixed -
Many of them are on migrations that we are not supposed to modify, or regarding a few characters over the per line limit by a comment, or defy styling conventions taught to us by LA.

File: spec/features/user_creates_account_spec.rb [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+  end
+
+  scenario 'user creates account' do
+    visit '/users/sign_up'
+
+    fill_in "Email", with: "derpherp@email.com"
+    fill_in "Password", with: "derpherp"
+    fill_in "Password confirmation", with: "derpherp"
+
+    click_on "Sign Up"
```
Comment (smkopp92) [inline spec/features/user_creates_account_spec.rb]: You should have an expectation here for the result of this scenario!

**Pull request merged**